### PR TITLE
 Write marshalled kubeletconfig object to init-config-dir

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
from @luxas :
>Write the the marshalled kubeletconfig object to /var/lib/kubelet/config/init/kubelet so that the kubelet will start up with the right params on init/join. The only params expected in the kubelet command-line after this is kubelet --init-config-dir /var/lib/kubelet/config/init --dynamic-config-dir /var/lib/kubelet/config/dynamic

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
ref: https://github.com/kubernetes/kubeadm/issues/28#issuecomment-345502933

**Special notes for your reviewer**:
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/constants/constants.go [hunk 1]
```diff
@@ -155,6 +155,12 @@ const (
 	// after https://github.com/kubernetes/kubernetes/pull/53833 being merged.
 	KubeletBaseConfigurationConfigMapKey = "kubelet"
 
+	// KubeletBaseConfigurationDir specifies the directory on the node where stores the initial remote configuration of kubelet
+	KubeletBaseConfigurationDir = "/var/lib/kubelet/config/init"
+
+	// KubeletBaseConfigurationFile specifies the file name on the node which stores initial remote configuration of kubelet
+	KubeletBaseConfigurationFile = "kubelet.yaml"
```
Comment (luxas) [inline cmd/kubeadm/app/constants/constants.go]: This constant is defined somewhere in the kubelet codebase. Right now it is `kubelet` I think. You can check the PR where @mtaufen is in process to graduate file-loading from alpha to beta

File: cmd/kubeadm/app/phases/kubelet/kubelet.go [hunk 2]
```diff
@@ -86,6 +90,15 @@ func UpdateNodeWithConfigMap(client clientset.Interface, nodeName string) error
 			return false, nil
 		}
 
+		baseCongfig, err := yaml.Marshal(kubeletCfg.Data[kubeadmconstants.KubeletBaseConfigurationConfigMapKey])
```
Comment (luxas) [inline cmd/kubeadm/app/phases/kubelet/kubelet.go]: I think you misunderstood this. The configuration that should be written to the file is `cfg.KubeletConfiguration.BaseConfig`; basically write the same thing you have in the configmap on disk in the init code.

In the join code, this config should be looked up from the configmap (we need a new RBAC rule for that)

Comment (xiangpengzhao) [reply cmd/kubeadm/app/phases/kubelet/kubelet.go]: The `kubeletCfg` here is the configmap we stored in API server, it's `Data[kubeadmconstants.KubeletBaseConfigurationConfigMapKey]` is the BaseConfig which we want.

See Line 50:
```go
kubeletBytes, err := kubeadmutil.MarshalToYamlForCodecs(cfg.KubeletConfiguration.BaseConfig, kubeletconfigv1alpha1.SchemeGroupVersion, *kubeletCodecs)
```
and Line 60-61:
```go
 		Data: map[string]string{
 			kubeadmconstants.KubeletBaseConfigurationConfigMapKey: string(kubeletBytes),
 		},
```

Feel free to correct me if I'm indeed misunderstood here :)

File: cmd/kubeadm/app/constants/constants.go (hunk 1 above)
Comment (xiangpengzhao) [reply cmd/kubeadm/app/constants/constants.go]: OK, I will check it. Thanks for pointing it out!

**Comment by xiangpengzhao** [comment]
bazel test failed. Seems like there is no write privilege to `/var/lib/kubelet/` ? @luxas any ideas?

```
//cmd/kubeadm/app/phases/kubelet:go_default_test 0.00s
bazel test //cmd/kubeadm/app/phases/kubelet:go_default_test
--- FAIL: TestCreateBaseKubeletConfiguration (0.52s)
	kubelet_test.go:83: CreateBaseKubeletConfiguration: unexepected error failed to write initial remote configuration of kubelet into file "/var/lib/kubelet/config/init/kubelet.yaml" for node fake-node: open /var/lib/kubelet/config/init/kubelet.yaml: no such file or directory
--- FAIL: TestUpdateNodeWithConfigMap (0.50s)
	kubelet_test.go:118: UpdateNodeWithConfigMap: unexepected error failed to write initial remote configuration of kubelet into file "/var/lib/kubelet/config/init/kubelet.yaml" for node fake-node: open /var/lib/kubelet/config/init/kubelet.yaml: no such file or directory
FAIL
```

File: cmd/kubeadm/app/phases/kubelet/kubelet.go (hunk 2 above)
Comment (xiangpengzhao) [reply cmd/kubeadm/app/phases/kubelet/kubelet.go]: >we need a new RBAC rule for that

What RBAC rule is needed? The rules for group "system:nodes" have already been created in init (`CreateBaseKubeletConfiguration`). Am I missing sth here?
```go
 	if err := createKubeletBaseConfigMapRBACRules(client); err != nil {
 		return fmt.Errorf("error creating base kubelet configmap RBAC rules: %v", err)
 	}
```

File: cmd/kubeadm/app/constants/constants.go (hunk 1 above)
Comment (xiangpengzhao) [reply cmd/kubeadm/app/constants/constants.go]: Updated. I add a TODO here.

https://github.com/kubernetes/kubernetes/pull/53833/files#diff-fbd96764b3922031c77bb5efdab3269aR21

File: cmd/kubeadm/app/phases/kubelet/kubelet.go (hunk 2 above)
Comment (mtaufen) [reply cmd/kubeadm/app/phases/kubelet/kubelet.go]: wrt RBAC, see the instructions regarding permissions here: https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/

Comment (xiangpengzhao) [reply cmd/kubeadm/app/phases/kubelet/kubelet.go]: Does that mean we should create separate rules for each node?

Comment (luxas) [reply cmd/kubeadm/app/phases/kubelet/kubelet.go]: Ok, I'll clarify this and alter my earlier plan a bit to make it better.
Let me know if this sounds good to you

1) There should be two functions: one that takes a `string` to write to disk (`WriteInitKubeletConfigToDisk`) and one for `kubeadm join` that takes a k8s client which extracts the config string (`GetKubeletConfigFromConfigMap`)
2) `WriteInitKubeletConfigToDisk` should:
 - Write the string down to this file `filepath.Join(KubeletBaseConfigurationDir, KubeletBaseConfigurationFile)`
3) `GetKubeletConfigFromConfigMap` should:
 - Get the base-config ConfigMap from the API server using a client that is passed in
 - Extract the value string from the `kubeadmconstants.KubeletBaseConfigurationConfigMapKey`

No `yaml.Marshal` or similar code should be used here.

**Review by luxas — COMMENTED** [review]
@xiangpengzhao The overall flow looks like this for `kubeadm join`:

```
1) token discovery as usual => `/etc/kubernetes/bootstrap-kubelet.conf` that will be used for TLS bootstrapping
2) get the base kubelet config ConfigMap using the bootstrap token as the credential
3) write the init config down to disk so the kubelet can start
4) The kubelet starts and performs the TLS bootstrap
5) kubeadm PATCHes the node to start using DKC using the kubelet's TLS bootstrapped credential
```

In `kubeadm init` this will happen:
```
... normal tasks like certs, kubeconfig, static pods, etc
1, NEW) Write the marshalled kubeletconfig down to the init config location
... now the kubelet can start the control plane
... control plane is up 
2) Create the kubeletconfig ConfigMap
3, EDIT) Allow `system:nodes` AND `system:bootstrappers:kubeadm:default-node-token` to access the ConfigMap. This is so `kubeadm join` can get the init config in step 2) above.
4) PATCH the master node to use Dynamic Kubelet Config
```

**Comment by xiangpengzhao** [comment]
@luxas thanks for clarification! updated. Hope this is close to what you expected :)

File: cmd/kubeadm/app/phases/kubelet/kubelet.go [hunk 3]
```diff
@@ -51,6 +56,10 @@ func CreateBaseKubeletConfiguration(cfg *kubeadmapi.MasterConfiguration, client
 		return err
 	}
 
+	if err := writeInitKubeletConfigToDisk(kubeletBytes); err != nil {
```
Comment (luxas) [inline cmd/kubeadm/app/phases/kubelet/kubelet.go]: This won't work. You need to write down this config RIGHT AFTER the controlplane/etcd phases in order for the kubelet to even start. So please extract this logic into a function you can call in `init.go` after the controlplane stuff, but before you start waiting for the API to be available

File: cmd/kubeadm/app/phases/kubelet/kubelet.go [hunk 4]
```diff
@@ -152,5 +180,51 @@ func createKubeletBaseConfigMapRBACRules(client clientset.Interface) error {
...
+		},
+		RoleRef: rbac.RoleRef{
+			APIGroup: rbac.GroupName,
+			Kind:     "ClusterRole",
+			Name:     tokenphase.NodeBootstrapperClusterRoleName,
+		},
+		Subjects: []rbac.Subject{
+			{
+				Kind: rbac.GroupKind,
+				Name: kubeadmconstants.NodeBootstrapTokenAuthGroup,
```
Comment (luxas) [inline cmd/kubeadm/app/phases/kubelet/kubelet.go]: Thanks :+1:

**Review by luxas — COMMENTED** [review]
Thanks! The only thing that needs fixing yet is that writing down the init config has to happen before waiting on the kubelet to launch the Static Pods, otherwise this will deadlock. But after you've re-organized this to write down the config after the controlplane phase, this should work well :)

One or two unit tests here wouldn't hurt though ;)

**Comment by luxas** [comment]
Please fix unit failure: `//cmd/kubeadm/app/phases/kubelet:go_default_test 0.00s`

File: cmd/kubeadm/app/phases/kubelet/kubelet.go [hunk 5]
```diff
@@ -152,5 +180,51 @@ func createKubeletBaseConfigMapRBACRules(client clientset.Interface) error {
 				Name: kubeadmconstants.NodesGroup,
 			},
 		},
+	}); err != nil {
+		return err
+	}
+
+	return apiclient.CreateOrUpdateClusterRoleBinding(client, &rbac.ClusterRoleBinding{
```
Comment (luxas) [inline cmd/kubeadm/app/phases/kubelet/kubelet.go]: Everything here can be just switched for one more subject in the above rolebinding -- just add the subject here to the RoleBinding above and it should work. That should grant `kubeadm join` using the token the right to get the configmap, right?

**Comment by xiangpengzhao** [comment]
comments addressed. unit test not updated yet :)

**Comment by xiangpengzhao** [comment]
/test pull-kubernetes-e2e-gce

@luxas I have to be off my keyboard for hours. Will update later.Can we catch up the 1.9 train?

File: cmd/kubeadm/app/phases/kubelet/kubelet.go [hunk 6]
```diff
@@ -148,9 +171,61 @@ func createKubeletBaseConfigMapRBACRules(client clientset.Interface) error {
...
+
+	if err := writeInitKubeletConfigToDisk(kubeletBytes); err != nil {
+		return fmt.Errorf("failed to write base configuration of kubelet to disk on master node %s: %v", cfg.NodeName, err)
+	}
+
+	return nil
+}
+
+func writeInitKubeletConfigToDisk(kubeletConfig []byte) error {
+	baseCongfigFile := filepath.Join(kubeadmconstants.KubeletBaseConfigurationDir, kubeadmconstants.KubeletBaseConfigurationFile)
```
Comment (luxas) [inline cmd/kubeadm/app/phases/kubelet/kubelet.go]: nit: `baseConfigFile`

File: cmd/kubeadm/app/phases/kubelet/kubelet.go [hunk 7]
```diff
@@ -148,9 +171,61 @@ func createKubeletBaseConfigMapRBACRules(client clientset.Interface) error {
...
+	if err := ioutil.WriteFile(baseCongfigFile, kubeletConfig, 0644); err != nil {
```
Comment (luxas) [inline cmd/kubeadm/app/phases/kubelet/kubelet.go]: you should create the directory here with `os.MkdirAll()` before trying to write the file

**Review by luxas — APPROVED** [review]
/lgtm
Two small-ish nits, but I verified that this works well locally as well :+1:

**Comment by xiangpengzhao** [comment]
Addressed comments, especially https://github.com/kubernetes/kubernetes/pull/56020#discussion_r152578255 in case it cause issues. @luxas PTAL. Thanks!

**Review by luxas — APPROVED** [review]
/lgtm

**Pull request merged**